### ISSUE TITLE
Automatically release memory

### ISSUE THREAD

**Issue by ttsite** [xmrig/xmrig-proxy#542]
Can you directly add the automatic memory release function on the agent end?

**Comment by SChernykh** [comment]
Can you explain more? xmrig-proxy doesn't use much memory to begin with.

**Comment by SChernykh** [comment]
How much does it increase over time? Is it constantly "leaking"? Then this is a bug, it shouldn't increase.

**Comment by ttsite** [comment]
> How much does it increase over time? Is it constantly "leaking"? Then this is a bug, it shouldn't increase.

Continuously increasing!

**Comment by SChernykh** [comment]
I'll look into it next week then. It shouldn't constantly increase. Which xmrig-proxy version do you use? One of the release binaries, or do you compile it yourself?

**Comment by SChernykh** [comment]
I did a quick test with https://github.com/google/sanitizers/wiki/AddressSanitizerLeakSanitizer and didn't find memory leaks. Do you use release binaries or do you compile xmrig-proxy yourself?

**Comment by ttsite** [comment]
> I did a quick test with https://github.com/google/sanitizers/wiki/AddressSanitizerLeakSanitizer and didn't find memory leaks. Do you use release binaries or do you compile xmrig-proxy yourself?

Both binary and self compiled tests have been conducted, which means that the memory usage will gradually increase with the increase of usage time. The version is the latest version, and I have been using third-party memory tools to organize it. My suggestion is whether the software can add a memory organizing function. Set a memory usage, and if it exceeds this usage, it will automatically organize and use memory to release it

**Comment by SChernykh** [comment]
"Memory organizing function" is not how C++ programs work. If memory usage is growing, it's a memory leak and it's a bug. Your OS can already automatically reduce memory used by programs (it's called swapping), no support from xmrig-proxy is needed.

**Comment by SChernykh** [comment]
Did you try values from this article https://dzone.com/articles/tcp-out-of-memory-consider-tuning-tcp-mem ?
```
net.core.netdev_max_backlog=30000
net.core.rmem_max=134217728
net.core.wmem_max=134217728
net.ipv4.tcp_max_syn_backlog=8192
net.ipv4.tcp_rmem=4096 87380 67108864
net.ipv4.tcp_wmem=4096 87380 67108864
```
And also `net.ipv4.tcp_mem = 4096 87380 67108864`

**Comment by bwq90** [comment]
Thanks for prompt response @SChernykh . I have already added these lines in my sysctl conf.

The lsof command just hangs on the server. probably due to high number of open files? :)

**Comment by bwq90** [comment]
in dmesg -T output. I see too many orphaned sockets output after xmrig is crashed.

```
[Tue Mar 26 10:41:04 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:04 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:04 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:04 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:04 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:08 2024] net_ratelimit: 39 callbacks suppressed
[Tue Mar 26 10:41:08 2024] TCP: out of memory -- consider tuning tcp_mem
[Tue Mar 26 10:41:08 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:08 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:08 2024] TCP: out of memory -- consider tuning tcp_mem
[Tue Mar 26 10:41:10 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:10 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:10 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:10 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:10 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:12 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:15 2024] net_ratelimit: 2 callbacks suppressed
[Tue Mar 26 10:41:15 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:15 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:17 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:17 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:17 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:17 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:17 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:17 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:17 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:17 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:21 2024] net_ratelimit: 1 callbacks suppressed
[Tue Mar 26 10:41:21 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:21 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:21 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:21 2024] TCP: too many orphaned sockets
[Tue Mar 26 10:41:21 2024] TCP: too many orphaned sockets
```

**Comment by bwq90** [comment]
Just to add. I have workers set to false already. But I have custom-diff set to 1000 and    

    "custom-diff": 1000,
    "custom-diff-stats": true,

Can this may have some impact due to high miner count?

**Comment by SChernykh** [comment]
Difficulty 1000 is too low, you'll be getting too many submitted shares and too high network load. XMRig donation servers set difficulty 1,000,000 for a reason.

**Comment by bwq90** [comment]
Thanks. What should be optimal setting for custom diff if I have  huge count of miners with different hardware CPU

**Comment by SChernykh** [comment]
It doesn't matter if a single miner has low hashrate and can't submit shares every 30 seconds. What matters is the overall load on the proxy, so you can just set difficulty = your total hashrate and get 1 incoming share/second on average.

**Comment by bwq90** [comment]
Thanks for the insight. I have changed it to 
```
"custom-diff": 0,
"custom-diff-stats": false,
```

I'll monitor for crash logs again. The Hardware of Proxy is top noch.

Intel(R) Xeon(R) CPU E5-2620 v3
128G RAM
SSD Disk

sysctl is tuned along with other Kernel tunings.

**Comment by bwq90** [comment]
Sorry for being a pain @SChernykh . One last concern. My only reason for lowering the custom diff to 1000 was that my miners are often offline and online. So what if they are assigned a high diff job and in the process of calculating the share. if They get offline and shutdown or timed out. In that scenario where there are a lof of such miners. Shouldn't we low the diff to lowest value so that we can get maximum work from miners when they are online?

**Comment by SChernykh** [comment]
No, this is not how mining works. Otherwise no one would be able to mine a block because no one is able to submit a 300G difficulty shares within a few hours they're online. Finding a share is a random memory-less process, and law of big numbers applies here. Many big miners = 1 big miner with the same hashrate for all practical purposes.

**Comment by bwq90** [comment]
Thank you so much. I have made the changes as per your suggestion and will monitor it for few days to see if proxy crashes for high memory. I hope it should not crash now. Secondly when should I decide that now is that time to add a parallel load balancer server with proxy?  What should be the maximum / optimal number of workers for a single Ubuntu Server with such specs?

Intel(R) Xeon(R) CPU E5-2620 v3
128G RAM
SSD Disk
